platform/linux-generic/pktio/mdev/e1000e.c

```diff
@@ -0,0 +1,605 @@
+		return -1;
+	}
+
+	txq->cidx =
+	    (odp_u32le_t *)(void *)(pkt_e1000e->mmio + E1000_TDH_OFFSET);
+
+	txq->tx_data.size = txq->tx_queue_len * E1000E_TX_BUF_SIZE;
+	ret = mdev_dma_area_alloc(&pkt_e1000e->mdev, &txq->tx_data);
+	if (ret) {
+		ODP_ERR("Cannot allocate TX queue DMA area\n");
+		return -1;
+	}
+
+	ODP_DBG("Register TX queue region: 0x%llx@%016llx\n", size, offset);
+	ODP_DBG("    TX descriptors: %u\n", txq->tx_queue_len);
+
+	return 0;
+}
+
+static int e1000e_region_info_cb(mdev_device_t *mdev,
+				 struct vfio_region_info *region_info)
+{
+	pktio_ops_e1000e_data_t *pkt_e1000e =
+	    odp_container_of(mdev, pktio_ops_e1000e_data_t, mdev);
+	mdev_region_class_t class_info;
+
+	if (vfio_get_region_cap_type(region_info, &class_info) < 0) {
+		ODP_ERR("Cannot find class_info in region %u\n",
+			region_info->index);
+		return -1;
+	}
+
+	switch (class_info.type) {
+	case VFIO_NET_MDEV_MMIO:
+		return e1000e_mmio_register(pkt_e1000e,
+					   region_info->offset,
+					   region_info->size);
+
+	case VFIO_NET_MDEV_RX_RING:
+		return e1000e_rx_queue_register(pkt_e1000e,
+					       region_info->offset,
+					       region_info->size);
+
+	case VFIO_NET_MDEV_TX_RING:
+		return e1000e_tx_queue_register(pkt_e1000e,
+					       region_info->offset,
+					       region_info->size);
+
+	default:
+		ODP_ERR("Unexpected region %u (class %u:%u)\n",
+			region_info->index, class_info.type,
+			class_info.subtype);
+		return -1;
+	}
+}
+
+static int e1000e_open(odp_pktio_t id ODP_UNUSED,
+		       pktio_entry_t *pktio_entry,
+		       const char *resource, odp_pool_t pool)
+{
+	pktio_ops_e1000e_data_t *pkt_e1000e;
+	int ret;
+
+	ODP_ASSERT(pool != ODP_POOL_INVALID);
+
+	if (strncmp(resource, NET_MDEV_PREFIX, strlen(NET_MDEV_PREFIX)))
+		return -1;
+
+	ODP_DBG("%s: probing resource %s\n", MODULE_NAME, resource);
+
+	pkt_e1000e = ODP_OPS_DATA_ALLOC(sizeof(*pkt_e1000e));
+	if (odp_unlikely(pkt_e1000e == NULL)) {
+		ODP_ERR("Failed to allocate pktio_ops_e1000e_data_t struct");
+		return -1;
+	}
+	pktio_entry->s.ops_data = pkt_e1000e;
+
+	memset(pkt_e1000e, 0, sizeof(*pkt_e1000e));
+
+	pkt_e1000e->sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (pkt_e1000e->sockfd == -1) {
+		ODP_ERR("Cannot get device control socket\n");
+		goto out;
+	}
+
+	ret =
+	    mdev_device_create(&pkt_e1000e->mdev, MODULE_NAME,
+			       resource + strlen(NET_MDEV_PREFIX),
+			       e1000e_region_info_cb);
+	if (ret)
+		goto out;
+
+	pkt_e1000e->pool = pool;
+
+	e1000e_wait_link_up(pktio_entry);
+
+	ODP_DBG("%s: open %s is successful\n", MODULE_NAME,
+		pkt_e1000e->mdev.if_name);
+
+	return 0;
+
+out:
+	e1000e_close(pktio_entry);
+	return -1;
+}
+
+static int e1000e_close(pktio_entry_t *pktio_entry)
+{
+	pktio_ops_e1000e_data_t *pkt_e1000e = pktio_entry->s.ops_data;
+
+	ODP_DBG("%s: close %s\n", MODULE_NAME, pkt_e1000e->mdev.if_name);
+
+	mdev_device_destroy(&pkt_e1000e->mdev);
+
+	for (uint16_t i = 0; i < pkt_e1000e->capa.max_input_queues; i++) {
```

Comment:
place type out of for(). checkpatch should warn about that.

Reply:
@muvarov This idiom is part of C99, so I see no problem with it here. It's also why checkpatch doesn't complain.

Reply:
./doc/application-api-guide/api_guide_lines.dox line 162

Reply:
@muvarov In this case the variable is being declared at the beginning of scope. The scope in this instance is the block defined by the for loop. That citation is saying you shouldn't declare variables in the middle of a scope. While that's still correct C, it can be confusing.

Reply:
Done.